From ab4d86d39c32c8e594c5f5d57b1342e02ca67931 Mon Sep 17 00:00:00 2001
From: Andy Wilkinson <awilkinson@pivotal.io>
Date: Fri, 23 Oct 2015 10:26:02 +0100
Subject: [PATCH] Update SmokeTests to be compatible with Spring Integration
 4.2

Spring Integration 4.2 has fixed a bug where integration components
were being started too early. SmokeTests was relying on this bug to
start the test class which is also a @MessageEndpoint.

This commit updates the test to use a nested inner-class for the
endpoint and a bean declaration in the XML configuration. This allows
the test to pass against both Spring Integration 4.0 and 4.2.
---
 .../batch/integration/SmokeTests.java              | 33 +++++++++++++---------
 .../batch/integration/SmokeTests-context.xml       |  1 +
 2 files changed, 20 insertions(+), 14 deletions(-)

diff --git a/spring-batch-integration/src/test/java/org/springframework/batch/integration/SmokeTests.java b/spring-batch-integration/src/test/java/org/springframework/batch/integration/SmokeTests.java
index bb5d102..1ac30d9 100644
--- a/spring-batch-integration/src/test/java/org/springframework/batch/integration/SmokeTests.java
+++ b/spring-batch-integration/src/test/java/org/springframework/batch/integration/SmokeTests.java
@@ -17,7 +17,6 @@ import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
 
 @ContextConfiguration
 @RunWith(SpringJUnit4ClassRunner.class)
-@MessageEndpoint
 public class SmokeTests {
 
 	@Autowired
@@ -26,18 +25,7 @@ public class SmokeTests {
 	@Autowired
 	private PollableChannel smokeout;
 
-	// This has to be static because Spring Integration registers the handler
-	// more than once (every time a test instance is created), but only one of
-	// them will get the message.
-	private volatile static int count = 0;
-
-	@ServiceActivator(inputChannel = "smokein", outputChannel = "smokeout")
-	public String process(String message) {
-		count++;
-		String result = message + ": " + count;
-		return result;
-	}
-
+
 	@Test
 	public void testDummyWithSimpleAssert() throws Exception {
 		assertTrue(true);
@@ -50,7 +38,24 @@ public class SmokeTests {
 		Message<String> message = (Message<String>) smokeout.receive(100);
 		String result = message == null ? null : message.getPayload();
 		assertEquals("foo: 1", result);
-		assertEquals(1, count);
+		assertEquals(1, AnnotatedEndpoint.count);
+	}
+
+	@MessageEndpoint
+	static class AnnotatedEndpoint {
+		
+		// This has to be static because Spring Integration registers the handler
+		// more than once (every time a test instance is created), but only one of
+		// them will get the message.
+		private volatile static int count = 0;
+		
+		@ServiceActivator(inputChannel = "smokein", outputChannel = "smokeout")
+		public String process(String message) {
+			count++;
+			String result = message + ": " + count;
+			return result;
+		}
+
 	}
 
 }
diff --git a/spring-batch-integration/src/test/resources/org/springframework/batch/integration/SmokeTests-context.xml b/spring-batch-integration/src/test/resources/org/springframework/batch/integration/SmokeTests-context.xml
index 292a06a..543ae48 100644
--- a/spring-batch-integration/src/test/resources/org/springframework/batch/integration/SmokeTests-context.xml
+++ b/spring-batch-integration/src/test/resources/org/springframework/batch/integration/SmokeTests-context.xml
@@ -12,6 +12,7 @@
                                  http://www.springframework.org/schema/tx/spring-tx-2.5.xsd">
                         
 	<annotation-config/>
+	<beans:bean class="org.springframework.batch.integration.SmokeTests.AnnotatedEndpoint"/>
 	<channel id="smokein"/>
 	<channel id="smokeout">
 		<queue/>
-- 
2.2.1

